packages/cosmic-swingset/src/launch-chain.js

```diff
@@ -733,6 +742,63 @@ export async function launch({
         return true;
       }
 
+      case ActionType.ENACTED_UPGRADE: {
+        // Install and execute new core proposals.
+        const {
+          upgradePlan: { info: upgradeInfoJson = null } = {},
+          blockHeight,
+          blockTime,
+        } = action;
+        const upgradePlanInfo = upgradeInfoJson && JSON.parse(upgradeInfoJson);
+
+        // Handle the planned core proposals as just another action.
+        const { coreProposals = [] } = upgradePlanInfo || {};
+
+        if (!coreProposals.length) {
+          // Nothing to do.
+          return undefined;
+        }
+
+        const { bundles, code: coreEvalCode } =
+          await extractCoreProposalBundles(coreProposals, vatconfig, {
+            handleToBundleSpec: async (handle, source, _sequence, _piece) => {
+              const bundle = await bundleSource(source);
+              const { endoZipBase64Sha512: hash } = bundle;
+              const bundleID = `b1-${hash}`;
+              handle.bundleID = bundleID;
+              harden(handle);
+              return harden([handle.bundleID, bundle]);
+            },
+          });
+
+        for (const bundle of Object.values(bundles)) {
+          await controller.validateAndInstallBundle(bundle);
```

Comment:
Should we handle errors?
Why not delegate to `installBundle` ?

Reply:
> Should we handle errors?

No, we should crash and propagate the rejection as usual.  Core proposals are trusted code, so we should be crisp when they fail.

> Why not delegate to `installBundle`?

`installBundle` is meant to be used by untrusted clients, so it seems too loose for this usage.  It takes a JSON blob as input, publishes to chain storage, and doesn't propagate errors very clearly.